contracts/bridge/contracts/common/fee-market/FeeMarket.sol

```diff
@@ -0,0 +1,416 @@
+pragma solidity >=0.6.0 <0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../../interfaces/IFeeMarket.sol";
+
+contract FeeMarket is IFeeMarket {
+    event SetOwner(address indexed o, address indexed n);
+    event SetOutbound(address indexed out, uint256 flag);
+    event Slash(address indexed src, uint wad);
+    event Reward(address indexed dst, uint wad);
+    event Deposit(address indexed dst, uint wad);
+    event Withdrawal(address indexed src, uint wad);
+    event Locked(address indexed src, uint wad);
+    event UnLocked(address indexed src, uint wad);
+    event AddRelayer(address indexed prev, address indexed cur, uint fee);
+    event RemoveRelayer(address indexed prev, address indexed cur);
+    event OrderAssgigned(uint256 indexed key, uint timestamp);
+    event OrderSettled(uint256 indexed key, uint timestamp);
+
+    address private constant SENTINEL_HEAD = address(0x1);
+    address private constant SENTINEL_TAIL = address(0x2);
+
+    // System treasury
+    address public immutable VAULT;
+    // The collateral relayer need to lock for each order.
+    uint256 public immutable COLLATERAL_PERORDER;
+    // Fee market assigned relayers numbers
+    uint256 public immutable ASSIGNED_RELAYERS_NUMBER;
+    // SLASH_AMOUNT = COLLATERAL_PERORDER * late_time / SLASH_TIME
+    uint256 public immutable SLASH_TIME;
+    // Time assigned relayer to relay messages
+    uint256 public immutable RELAY_TIME;
```

Comment:
Configurable by owner, and record locked_collateral and assigned_relayer_count in order.